server/src/main/java/org/opensearch/rest/action/cat/RestNodesAction.java

```diff
@@ -120,49 +130,177 @@ public RestChannelConsumer doCatRequest(final RestRequest request, final NodeCli
-                nodesInfoRequest.setIncludeDiscoveryNodes(false);
-                nodesInfoRequest.clear()
-                    .addMetrics(
-                        NodesInfoRequest.Metric.JVM.metricName(),
-                        NodesInfoRequest.Metric.OS.metricName(),
-                        NodesInfoRequest.Metric.PROCESS.metricName(),
-                        NodesInfoRequest.Metric.HTTP.metricName()
+                long leftTimeNano = timeoutNano - System.nanoTime() + beginTimeNano;
+                if (leftTimeNano < TIMEOUT_THRESHOLD_NANO) {
+                    onFailure(
+                        new OpenSearchTimeoutException(
+                            "There is not enough time to obtain nodesInfo metric from the cluster manager:"
+                                + clusterStateResponse.getState().nodes().getMasterNode().getName()
+                        )
                     );
-                client.admin().cluster().nodesInfo(nodesInfoRequest, new RestActionListener<NodesInfoResponse>(channel) {
-                    @Override
-                    public void processResponse(final NodesInfoResponse nodesInfoResponse) {
-                        NodesStatsRequest nodesStatsRequest = new NodesStatsRequest();
-                        nodesStatsRequest.timeout(request.param("timeout"));
-                        nodesStatsRequest.setIncludeDiscoveryNodes(false);
-                        nodesStatsRequest.clear()
-                            .indices(true)
-                            .addMetrics(
-                                NodesStatsRequest.Metric.JVM.metricName(),
-                                NodesStatsRequest.Metric.OS.metricName(),
-                                NodesStatsRequest.Metric.FS.metricName(),
-                                NodesStatsRequest.Metric.PROCESS.metricName(),
-                                NodesStatsRequest.Metric.SCRIPT.metricName()
+                    return;
+                }
+                String[] nodeIds = clusterStateResponse.getState().nodes().resolveNodes();
+                ConcurrentMap<String, NodeInfo> successNodeInfos = new ConcurrentHashMap<>(nodeIds.length);
+                ConcurrentMap<String, FailedNodeException> failNodeInfos = new ConcurrentHashMap<>(nodeIds.length);
+                ConcurrentMap<String, NodeStats> successNodeStats = new ConcurrentHashMap<>(nodeIds.length);
+                ConcurrentMap<String, FailedNodeException> failNodeStats = new ConcurrentHashMap<>(nodeIds.length);
+                AtomicInteger counter = new AtomicInteger();
+                for (String nodeId : nodeIds) {
+                    NodesInfoRequest nodesInfoRequest = createNodesInfoRequest(timeoutNano, leftTimeNano, nodeId);
+                    nodesInfoRequest.setIncludeDiscoveryNodes(false);
+                    client.admin().cluster().nodesInfo(nodesInfoRequest, new ActionListener<>() {
+                        @Override
```

Comment:
@aasom143, Iterator over each nodeIds can solve the two problems:
1. If timeout from the client is 30s, without adding `cluster().state()`, the overall time can be 60s, which is 2x times that the expect.
2. Only all nodes return the NodeInfoResponse in `cluster().nodesInfo()`, can the next `cluster().nodesStats()` be called. It's normal to have a slow node(such as fullGc) in large clusters, the api will become unresponsive, it means that if some of nodes are blocked in `cluster().nodesInfo()`, the api will be blocked.

`Also NodesInfo is not sync call, it's async`, can you explain it more clearly? The current implementation is all async.

Reply:
1. I agree with your point that the timeout can become 2X. 
2. Let's clarify with an example: if you have 100 nodes and the node info takes 10 seconds for each node, the total time will be 1000 seconds, since you are making a call for node info and waiting for a response, even if node info call is async. However, in the existing code, we make a broadcast call to each node within seconds, resulting in a total timeout of 10 seconds plus an additional 2 seconds for the broadcast call. Please let me know if I am incorrect.

Reply:
In the new logic, if the cluster has 100 nodes, and the node info takes 10 seconds for each node, As `cluster().nodesInfo()` is call async, the coordinator just send the `NodesInfoRequest` request to a node, and then processes the next node, `since you are making a call for node info and waiting for a response` is not right.

 The total time will be 10 seconds, an additional 2 seconds for the broadcast call in the new code.

Reply:
> since you are making a call for node info and waiting for a response is not right.

Under the for loop you are waiting for the response of nodes info
```
            public void onResponse(NodesInfoResponse nodesInfoResponse) {
                    assert nodesInfoResponse.getNodes().size() + nodesInfoResponse.failures().size() == 1;
```

So until you get the response it won't go to next node.